src/notebooks/controllers/preferredKernelConnectionService.node.ts

```diff
@@ -87,3 +77,35 @@ export function findPythonEnvBelongingToFolder(folder: Uri, pythonEnvs: readonly
         ? localEnvs[0]
         : undefined;
 }
+
+export async function getRecommendedPythonEnvironment(uri: Uri): Promise<ResolvedEnvironment | undefined> {
+    type RecommededEnvironment =
+        | {
+              environment: ResolvedEnvironment;
+              reason: 'globalUserSelected' | 'workspaceUserSelected' | 'defaultRecommended';
+          }
+        | undefined;
+    try {
+        const result = (await raceTimeout(5_000, commands.executeCommand('python.getRecommendedEnvironment', uri))) as
+            | RecommededEnvironment
+            | undefined;
+        if (!result) {
+            return;
+        }
+        if (result.reason === 'workspaceUserSelected') {
+            return result.environment;
+        }
+
+        // In dev containers give preference to the global user selected environment.
+        // This is because the global user selected environment is the one that is most likely pre-configured in the dev container.
+        // Generally the images have the settings already pre-configured with this global env pre-selected.
+        if (
+            result.reason === 'globalUserSelected' &&
+            (!workspace.workspaceFolders?.length || env.remoteName === 'dev-container')
```

Comment:
Is `!workspace.workspaceFolders?.length` a state that a dev container would be in, or is this covering another check as well? Seems new to this implementation

Reply:
Oh will update comments